boot/assets.go

```diff
@@ -20,27 +20,128 @@
 package boot
 
 import (
+	"crypto"
+	_ "crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
 	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/bootloader"
+	"github.com/snapcore/snapd/dirs"
 	"github.com/snapcore/snapd/gadget"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/strutil"
 )
 
+type trustedAssetsCache struct {
+	cacheDir string
+	hash     crypto.Hash
+}
+
+func newTrustedAssetsCache(cacheDir string) *trustedAssetsCache {
+	return &trustedAssetsCache{cacheDir: cacheDir, hash: crypto.SHA256}
+}
+
+func (c *trustedAssetsCache) assetKey(blName, assetName, assetHash string) string {
+	return filepath.Join(blName, fmt.Sprintf("%s-%s", assetName, assetHash))
+}
+
+func (c *trustedAssetsCache) tempAssetKey(blName, assetName string) string {
+	return filepath.Join(blName, assetName+".temp")
+}
+
+func (c *trustedAssetsCache) pathInCache(part string) string {
+	return filepath.Join(c.cacheDir, part)
+}
+
+// Add entry for
+func (c *trustedAssetsCache) Add(assetPath, blName, assetName string) (ta *trackedAsset, added bool, err error) {
```

Comment:
as I said elsewhere, except maybe for tests, added bool seems dangerous, also because anyway we can't update the cache and modeenv atomically at the same time, so there's always a chance that something is in the cache but wasn't added yet to modeenv